sycl/doc/extensions/proposed/sycl_ext_oneapi_joint_for.asciidoc

```diff
@@ -0,0 +1,261 @@
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2024 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+A very common idiom in SPMD programming sees developers distribute a range over
+the members of a group, by writing a loop that looks like this:
+
+[source,c++]
+----
+for (int i = it.get_local_id(0); i < N; i += it.get_local_range(0)) { ... }
+----
+
+The combination of setting the initial value based on the work-item index and
+setting the stride based on the number of work-items in the group ensures that
+any data accesses within the loop are appropriately interleaved to maximize
+performance.
+
+Although common, this idiom has several drawbacks: it is difficult to teach to
+developers unfamiliar with SPMD programming; it is easy to use incorrectly
+(e.g., by forgetting to set a different initial value on each work-item); and
+it is difficult to use generically (i.e., for different groups of work-items).
+
+This extension proposes to encapsulate this pattern in group algorithms, to
+simplify generic usage across different group types.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_JOINT_FOR` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Group algorithms
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <typename Group, typename InputIterator, typename Function>
+void joint_for_each(Group g, InputIterator first, InputIterator last, Function f);
```

Comment:
Why that and not something like
```
  for (auto &&elem : group_view_name(g, first, last)) /* ... */
```
?

Reply:
The rest of the functions in the group algorithms library are based on `std::` algorithms, (e.g., `std::reduce` becomes `sycl::joint_reduce`, `std::exclusive_scan` becomes `sycl::joint_exclusive_scan`) and so following that precedent here seems the most natural: `std::for_each` becomes `sycl::joint_for_each`.